### DISCUSSION TITLE
define submeshes for a new format

### DISCUSSION THREAD

**Discussion by MoiseRousseau** [nschloe/meshio#1196] [General]
Hi all!
I try to implement a new reader/writer for meshio to handle PFLOTRAN meshes (https://www.pflotran.org/). Such format can handle "submeshes" which define set of cells to further apply different material properties or boundary conditions.  However, I can't figure out in meshio how to store these information during reading or to use it during write. In fact, I see two mesh attributes (cell_sets and cell_tags) which can do the job,. Is it one of the two I must used ? If yes, how do we specify it ?
Have a nice day,
Moise

**Comment by nschloe** [comment]
Depends on how your information is stored. If sets are stored as arrays of indices into the cells array, `cell_sets` is for you. Some formats specify sets as functions over the cells, i.e., each cell has exactly one (integer) value associated with it. If this is case, you want to look at `cell_data`.

**Comment by MoiseRousseau** [comment]
It is stored as array of indices, so I would go for cell_sets. What do I must define then ? A list of each cell types ? For example, if:

cells = [
    ("triangle", [[0, 1, 2], [1, 3, 2]]),
    ("quad", [[1, 4, 5, 3]]),
] #like the example on meshio page

and if would create a cell_set with the second triangle and the first quad, i must enter:

cell_sets = {"cell_set name" : [("triangle", [1]), ("quad":[0]) ] }

 ?

Also, what the difference with a cell_tag like used for med reader ?

**Reply by nschloe** [reply]
That looks correct. 

> Also, what the difference with a cell_tag like used for med reader ?

I was going to say that `cell_tag` gets merged into `cell_data` eventually, but for med, it's different:
```python
mesh.point_tags = point_tags                                                                                                                                                              
mesh.cell_tags = cell_tags
```
I cannot remember why we do it like this, must be some workaround. It's definitely not in the style of meshio.

**Reply by MoiseRousseau** [reply]
Thanks for these clarifications. Everything clear now!
